Allow the use of Mbed Studio's version of ARMC6

bridadan opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->

This seems to be in good working order so far. I need to add a unit test before this is ready to go, hence raising this as a Draft pull request. Raising this early for coordination efforts and to get reviews started.

This allows people to use the ARMC6 compiler bundled with Mbed Studio with the command line tools. It also preserves the ability to use the professional compiler and license server.

FYI @ChiefBureaucraticOfficer @SenRamakri @arekzaluski @thegecko 

### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

You can use the ARMC6 compiler that is bundled with Mbed Studio with the command line tools. Enable this behavior by adding the ARMC6 executable to your system path (the default location of this on Windows is `C:/MbedStudio/tools/ac6/bin`). Alternatively, you can set this same location with Mbed CLI with the command `mbed config ARMC6_PATH C:/MbedStudio/tools/ac6/bin` (be sure to use the correct path when you run the command). You will also need to set the `ARMLMD_LICENSE_FILE` environment variable to the license file that is bundled with Mbed Studio. Check the Mbed OS documentation for further details.

arekzaluski reviewed (changes requested):
Building in Mbed Studio using this version of mbed-os-tools no longer works out of the box. Mbed Studio fails with an error:

`No valid license file found for the Mbed Studio version of ARMC6`

![Screen Shot 2019-03-18 at 15 36 05](https://user-images.githubusercontent.com/13417215/54542349-90cec180-4993-11e9-803e-5637db3eb95f.png)

Need to do more tests and check the code to understand why.

arekzaluski commented on tools/toolchains/arm.py:
```diff
@@ -561,25 +603,64 @@ def is_not_supported_error(self, output):
+                    "ARMLMD_LICENSE_FILE": license_file
+                }
+            else:
+                raise ToolException(
```
This code doesn't work when `ARMLMD_LICENSE_FILE` is passed. Else code is executed if license exists in different location than expected.

arekzaluski reviewed (changes requested):
Unfortunately, it still doesn't work for me in Mbed Studio.

I see following error while building:
[ERROR] armclang: error: Failed to checkout out a license. The license file could not be found. Check the ARMLMD_LICENSE_FILE is set correctly.

ARMLMD_LICENSE_FILE env variable is set to : `{MY_SYSTEM_PATH}/ac6-license.dat;{MY_SYSTEM_PATH}/ac6-license.dat`.

It seems that mbed os tools are adding the same path (correct one) twice. ARM Compiler 6 fails in that scenario.

arekzaluski commented on tools/toolchains/arm.py:
```diff
@@ -546,12 +580,21 @@ def __init__(self, target, *args, **kwargs):
 
+    @property
+    def is_mbed_studio_armc6(self):
+        return self.product_name and "Mbed Studio" in self.product_name
```
I'm not convinced that mbed os tools should have special logic based on a single product. Not sure if there is a better way of fixing it but it should be worth double checking before merging it.

arekzaluski replied:
It also limit options to make any changes in product name if needed. As old mbed os versions will be already released.

bridadan replied:
> I'm not convinced that mbed os tools should have special logic based on a single product. 

Considering the compiler has special behavior in this single product, I don't know if we have much of a choice 😄. The license shipped with Mbed Studio **must** be supplied to the compiler shipped with Mbed Studio in order to build anything. However if you have any alternatives/suggestions do let me know.

> It also limit options to make any changes in product name if needed. As old mbed os versions will be already released.

This is the only way I know of to identify this build of the compiler. As you can see, the output of `armclang --vsn` is quite similar between the different builds:

ARMC6 Professional:
```
Product: ARM Compiler 6.11 Professional
Component: ARM Compiler 6.11
Tool: armclang [5d3b4200]
```

ARMC6 from Mbed Studio:
```
<removed license error for brevity>
Product: ARM Compiler 6.11 for Mbed Studio
Component: ARM Compiler 6.11
Tool: armclang [5d3b3c00]
```

And we shouldn't be switching this behavoir based on the commit since that will change with new releases of the compiler.

thegecko replied:
I agree with @arekzaluski, managing specific cases like this is problematic and prone to breaking. For example, we plan to move the installation directory of the tools in the next release of Mbed Studio.
At a higher level, requiring Mbed Studio to be installed for this functionality to work makes little sense IMO.

bridadan replied:
@thegecko I understand your concern, it definitely reduces the flexibility of how Mbed Studio is internally organized.

> For example, we plan to move the installation directory of the tools in the next release of Mbed Studio.

This isn't immediately a problem for this implementation, as it only requires the `ac6-license.dat` file to be placed in any parent directory of the compiler binary.

> At a higher level, requiring Mbed Studio to be installed for this functionality to work makes little sense IMO.

Mbed Studio is only required to be installed if you wish to take advantage of the development license and corresponding compiler provided by Mbed Studio. You can of course still use the professional versions of the Arm Compiler 6 and license.

---

Another option is that I completely remove all of the "license detection" features at the expense of user experience. This would mean the user would now be responsible for the following:

1. Find the `ac6-license.dat` file in the Mbed Studio installation
1. Add the `ARMLMD_LICENSE_FILE` environment variable to their shell and set it to the path to the `ac6-license.dat` file.

We can of course document all of this, but that also means we need to keep this documentation up to date whenever the structure of Mbed Studio changes.

If the license detection feature is something that we need to ship though, I really need an alternative suggestion from your team. I'll happily parse a file to get the exact location. Maybe you could place a dedicated file next to the armc6 binaries? That file could point to the correct location of the license.

arekzaluski replied:
I was thinking about it and I think that the best way would be to remove license detection feature from mbed-os tools. Instead, it should be `applications` (Mbed Studio, mbed-cli, possible other in the future) that controls specific compilation behavior. Only they know location of all of the files.

At the moment Mbed Studio handles cases when `ARMLMD_LICENSE_FILE` env variable is set or/and AC6 compiler is installed on user's system. It simply overrides it and set its own path for both license and compiler location. It overrides it only for a single build process so user's environment variables are not changed.

I believe that the best way forward is to add similar logic in mbed-cli. For example installers of mbed-cli can be shipped with ARMC6 and valid license. Installer knows installation location so it can ask user at the end of installation if env variables should be updated or not.
Of course it should be also well documented for users that do not use installers to install mbed-cli.

@bridadan What do you think about it?

bridadan replied:
Seems like the least contentious way forward, so I'll go with that. Moving this functionality to the installers doesn't sound like a bad idea either 👍

cmonr replied:
@bridadan Checking since this PR has passed CI checks.

I'm assuming the above will be planned for 5.13, and _not_ squeezed into 5.12.0, correct?

bridadan replied:
Based on the feedback from the Mbed Studio folks it sounds like they are not comfortable with this PR going in as-is for 5.12, so I think I have to make the tool changes now. The installers will not be ready though for sometime.

bridadan replied:
For what its worth I have the changes almost ready. I don't have access to a mac to test though.

arekzaluski commented on tools/toolchains/arm.py:
```diff
@@ -429,6 +461,8 @@ class ARMC6(ARM_STD):
         "Cortex-A9"
     ]
     ARMCC_RANGE = (LooseVersion("6.10"), LooseVersion("7.0"))
+    MBED_STUDIO_LICENSE_FILE_NAME = "ac6-license.dat"
```
Same here. Hardcoding the name will also limit any options of changing the file name or its location. I think it is worth revisiting if there is any other option of doing it.

bridadan replied:
At some point the Mbed OS tools need to know what license file to pass to the compiler. So we need either a file to always pass to the compiler, or we would need some kind of file (in a location that doesn't move) that we could parse to then get the location of the license file/directory. Since 0.4.0 has already shipped without such a file (to my knowledge), I think we probably need to go with what's currently implemented. Though again, I'm open to suggestions!

0Grit replied:
Please consider for the future that if this version of ARMC6 and license are made available for use outside Mbed Studio our team will probably be doing so.

We ensure all our devs are using the same tool-chain by containerizing it.

bridadan committed:
Whitespace clean up

bridadan committed:
Conditionally enable --ide=mbed from ARMC6 based on compiler version.

If the Mbed Studio build of ARMC6 is found, add the --ide=mbed to all
necessary executables.

bridadan committed:
Remove stray prints and whitespace

bridadan committed:
Test detection of Mbed Studio version of ARMC6

bridadan committed:
Clean up some whitespace

bridadan committed:
Version check the compiler in all build functions

This enables the use of Mbed Studio's version of ARMC6.

thegecko commented on tools/toolchains/arm.py:
```diff
@@ -594,3 +640,21 @@ def compile(self, cc, source, object, includes):
+        )
+
+        if self.is_mbed_studio_armc6:
+            cmd.insert(1, "--ide=mbed")
```
IMO, this flag should be passed in whenever running AC6, not specifically when `is_mbed_studio_armc6`. The detection of Mbed Studio version can then be removed.

bridadan replied:
It was deemed necessary to do the conditional insertion of the `--ide=mbed` flag because supplying this to the professional version of ARMC6 causes a warning to show for every compile, assemble, and link step. The warning is along the lines of "unused argument --ide=mbed".

thegecko reviewed (approved):
LGTM

0xc0170 reviewed (dismissed):
Needs our final approval based on some recent concerns that are being taken care of = request changes.

bridadan committed:
Add comments about the proper use of specific ARMC6 arguments
